### PR TITLE
[database] Fix, tables endpoint

### PR DESCRIPTION
**Pull request by dpgaspar**
### CATEGORY

Choose one

- [x] Bug Fix
- [ ] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
Adds database filter to `/superset/tables`. Took the chance to add typing, improved parameter validation and tests

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS
@willbarrett

### CODE REVIEW

File: superset/views/core.py [hunk 1]
```diff
@@ -1068,30 +1054,47 @@ def schemas(self, db_id, force_refresh="false"):
...
-        db_id = int(db_id)
-        force_refresh = force_refresh.lower() == "true"
-        schema = utils.parse_js_uri_path_item(schema, eval_undefined=True)
-        substr = utils.parse_js_uri_path_item(substr, eval_undefined=True)
-        database = db.session.query(models.Database).filter_by(id=db_id).one()
+        logger.warning(
+            "/superset/tables/ API endpoint "
+            "is deprecated and will be removed in version 1.0.0"
+        )
+        stats_logger.incr(f"{self.__class__.__name__}.tables.init")
```
Comment (willbarrett) [inline superset/views/core.py]: `.init` seems odd - this isn't an initialization function. Is that copy/pasta or was it intentional? Should this include the `deprecated` prefix?

File: superset/views/core.py [hunk 2]
```diff
@@ -1105,20 +1108,22 @@ def tables(self, db_id, schema, substr, force_refresh="false"):
...
+            database, views, schema_parsed
         )
 
         def get_datasource_label(ds_name: utils.DatasourceName) -> str:
-            return ds_name.table if schema else f"{ds_name.schema}.{ds_name.table}"
+            return (
+                ds_name.table if schema_parsed else f"{ds_name.schema}.{ds_name.table}"
+            )
 
         if substr:
```
Comment (willbarrett) [inline superset/views/core.py]: Should this now be `if substr_parsed`?

File: superset/views/database/api.py [hunk 3]
```diff
@@ -274,6 +351,139 @@ def table_metadata(self, database: Database, table_name: str, schema_name: str):
...
+        )
+
+        if not schema_parsed and database.default_schemas:
+            user_schema = g.user.email.split("@")[0]
+            valid_schemas = set(database.default_schemas + [user_schema])
+
+            tables = [tn for tn in tables if tn.schema in valid_schemas]
+            views = [vn for vn in views if vn.schema in valid_schemas]
+
+        max_items = app.config["MAX_TABLE_NAMES"] or len(tables)
```
Comment (willbarrett) [inline superset/views/database/api.py]: Is there a way we can get at this configuration value without pulling in the entire `app` object? Is it possible to use the `current_app` proxy or load the config value directly?

File: superset/views/database/api.py [hunk 4]
```diff
@@ -274,6 +351,139 @@ def table_metadata(self, database: Database, table_name: str, schema_name: str):
...
 
+    @expose("/<int:pk>/tables/<string:schema_name>/<string:substr>/", methods=["GET"])
+    @expose(
+        "/<int:pk>/tables/<string:schema_name>/<string:substr>/<force_refresh>/",
+        methods=["GET"],
+    )
+    @protect()
+    @safe
+    @event_logger.log_this
+    def tables(  # pylint: disable=invalid-name,too-many-locals
```
Comment (willbarrett) [inline superset/views/database/api.py]: I'm starting to wonder if we want to do this migration now as part of the security fix PR? This looks like we've pulled the method definition over almost as-is, but as `pylint` is showing us, this really could use a refactor. Would it be better to have one PR that adds the security fix, then a second that refactors the method into the new API?

Another approach here could be to "strangle" the current implementation by proxying to the current implementation, then refactoring it over time.

**Review by willbarrett — CHANGES_REQUESTED** [review]
I think this should be 2 PRs - one implementing the security fix on the current endpoint, and a second migrating the endpoint to the new API. We're mixing concerns, and I think the migration to the new API will require a rework of the code.

File: superset/views/core.py (hunk 1 above)
Comment (dpgaspar) [reply superset/views/core.py]: Fixed, yes `init` may be debatable but was the metric name agreed on #9054

File: superset/views/core.py (hunk 2 above)
Comment (dpgaspar) [reply superset/views/core.py]: Yes, great catch!

File: superset/views/database/api.py (hunk 3 above)
Comment (dpgaspar) [reply superset/views/database/api.py]: Changed it to `current_app`

File: superset/views/database/api.py (hunk 4 above)
Comment (dpgaspar) [reply superset/views/database/api.py]: Agreed I'll break this into 2 PR's

File: superset/views/core.py [hunk 5]
```diff
@@ -1068,30 +1054,39 @@ def schemas(self, db_id, force_refresh="false"):
...
-        substr = utils.parse_js_uri_path_item(substr, eval_undefined=True)
-        database = db.session.query(models.Database).filter_by(id=db_id).one()
+        # Guarantees database filtering by security access
+        query = db.session.query(models.Database)
+        query = DatabaseFilter("id", SQLAInterface(models.Database, db.session)).apply(
+            query, None
+        )
+        database = query.filter_by(id=db_id).one_or_none()
+        if not database:
+            return json_error_response("Not found", 404)
```
Comment (villebro) [inline superset/views/core.py]: Did we settle for try-except with `.one()` or `.one_or_none()` with `if not`? I'm fine either way, just prefer to have it as uniform as possible in the codebase.

Comment (villebro) [inline superset/views/core.py]: Did we settle for try-except with `.one()` or `.one_or_none()` with `if not`? I'm fine either way, just prefer to have it as uniform as possible in the codebase.

**Review by villebro — APPROVED** [review]
I agree with @willbarrett , much better to have the original PR split up as separate PRs.

File: superset/views/core.py (hunk 5 above)
Comment (dpgaspar) [reply superset/views/core.py]: There is a majority of `one_or_none()` on `core.py`

**Pull request merged**